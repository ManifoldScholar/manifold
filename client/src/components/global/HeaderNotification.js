import React, { Component, PropTypes } from 'react';
import classNames from 'classnames';

export default class HeaderNotification extends Component {
  static propTypes = {
    id: PropTypes.string,
    heading: PropTypes.string,
    body: PropTypes.string,
    level: PropTypes.number,
    removeNotification: PropTypes.func
  };

  constructor(props) {
    super(props);
    this.handleClose = this.handleClose.bind(this);
  }

  // Close notification in handler in case event access is required
  handleClose() {
    this.props.removeNotification(this.props.id);
  }

  bodyCopy() {
    let output = null;
    if (this.props.body) {
      output = (
        <p className="notification-body">
          {this.props.body}
        </p>
      );
    }

    return output;
  }

  render() {
    const notificationClass = classNames({
      'header-notification': true,
      notice: this.props.level === 0,
      warning: this.props.level === 1,
      error: this.props.level === 2
    });
    return (
      <div className={notificationClass} key={this.props.id}>
        <div className="container">
          <header>
            <h5 className="notification-heading">{this.props.heading}</h5>
          </header>
          {this.bodyCopy()}

          <button className="notification-close" onClick={this.handleClose} data-id="close">
            <i className="manicon manicon-x"></i>
            <span className="screen-reader-text">{'Click to close this notification'}</span>
          </button>
        </div>
      </div>
    );
  }
}
