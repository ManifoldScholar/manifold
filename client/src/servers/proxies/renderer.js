import config from "config";
import proxy from "http-proxy-middleware";
import ch from "../../helpers/consoleHelpers";
import createStore from "../../store/createStore";
import exceptionRenderer from "../../helpers/exceptionRenderer";
import manifoldBootstrap from "bootstrap";
import CookieHelper from "helpers/cookie/Server";

const ssrRenderUrl = `http://${config.services.client.domain}:${config.services.client.sparePort}`;

export default function makeRendererProxy(requestHandler) {
  const reactServerProxy = proxy({
    target: ssrRenderUrl,
    changeOrigin: true,
    logLevel: "silent",

    onError: (err, req, res) => {
      ch.error(`Error proxying to port ${ssrRenderUrl}`);
      ch.error(`Perhaps the SSR service is in the process of reloading?`);
      ch.error(`Falling back to SSR rescue service only`);

      const store = createStore();

      const render = () => {
        try {
          requestHandler(req, res);
        } catch (error) {
          if (error.code === "MODULE_NOT_FOUND") {
            const msg =
              "Waiting for initial Webpack build to complete. Wait a few seconds " +
              "and reload.";
            ch.error(msg);
            res.setHeader("Content-Type", "text/html");
            res.end(msg);
          } else {
            ch.error(
              `Rendering fallback failed to render inn server-developmet.js`
            );
            res.setHeader("Content-Type", "text/html");
            res.end(exceptionRenderer(error));
          }
        }
      };

      const promises = [];
      const cookie = new CookieHelper(req, res);
      promises.push(manifoldBootstrap(store.getState, store.dispatch, cookie));
      Promise.all(promises).then(render, render);
    }
  });
  return reactServerProxy;
}
