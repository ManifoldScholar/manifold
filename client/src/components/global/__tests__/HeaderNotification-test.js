import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import HeaderNotification from '../HeaderNotification';

describe("Global.HeaderNotification component", () => {

  const removeMock = jest.fn();
  const root = (
    <HeaderNotification
      id="1"
      heading="Welcome to the Terrordome"
      body="Here's your ticket.  Every time I get wicked."
      level={2}
      removeNotification={removeMock}
    />
  );

  it('renders correctly', () => {
    const component = renderer.create(root);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should trigger removeNotification callback when close is clicked', () => {
    const wrapper = mount(root);
    removeMock.mockClear();
    wrapper.find('[data-id="close"]').first().simulate('click');
    expect(removeMock).toHaveBeenCalled();
  });

});

