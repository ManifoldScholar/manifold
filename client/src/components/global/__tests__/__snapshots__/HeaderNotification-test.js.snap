// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global.HeaderNotification component renders correctly 1`] = `
<div
  className="header-notification error"
>
  <div
    className="container"
  >
    <header>
      <h5
        className="notification-heading"
      >
        Welcome to the Terrordome
      </h5>
    </header>
    <p
      className="notification-body"
    >
      Here's your ticket.  Every time I get wicked.
    </p>
    <button
      className="notification-close"
      data-id="close"
      onClick={[Function]}
    >
      <i
        className="manicon manicon-x"
      />
      <span
        className="screen-reader-text"
      >
        Click to close this notification
      </span>
    </button>
  </div>
</div>
`;
